dhall/src/Dhall/Import.hs

```diff
@@ -1019,62 +1005,83 @@ getCacheBaseDirectory = alternative₀ <|> alternative₁ <|> alternative₂
+        },
+        srcEnd = SourcePos {
+            sourceName = fakeSrcName,
+            sourceLine = mkPos 1,
+            sourceColumn = mkPos (Text.length fakeSrcText)
+        },
+        srcText = fakeSrcText
+    }
+  where
+    fakeSrcText = "«Origin Header Configuration»"
+    fakeSrcName = "[builtin]"
+
+-- | Load headers only from the environment (used in tests)
+envOriginHeaders :: Expr Src Import
+envOriginHeaders = Note headersSrc (Embed (Import (ImportHashed Nothing (Env "DHALL_HEADERS")) OriginHeaders))
+
+-- | Load headers in env, falling back to config file
+defaultOriginHeaders :: IO (Expr Src Import)
+defaultOriginHeaders = do
+    fromFile <- originHeadersFileExpr
+    return (ImportAlt envOriginHeaders (Note headersSrc fromFile))
+
+-- | Given a headers expression, return an origin headers loader
+originHeadersLoader :: IO (Expr Src Import) -> StateT Status IO OriginHeaders
+originHeadersLoader headersExpr = do
+    partialExpr <- liftIO headersExpr
 
-    let go key₀ key₁ = do
-            let expected :: Expr Src Void
-                expected =
-                    App List
-                        ( Record $ Core.makeRecordField <$>
-                            Dhall.Map.fromList
-                                [ (key₀, Text)
-                                , (key₁, Text)
-                                ]
-                        )
-
-            let suffix_ = Dhall.Pretty.Internal.prettyToStrictText expected
-            let annot = case loadedExpr of
-                    Note (Src begin end bytes) _ ->
-                        Note (Src begin end bytes') (Annot loadedExpr expected)
-                      where
-                        bytes' = bytes <> " : " <> suffix_
-                    _ ->
-                        Annot loadedExpr expected
-
-            _ <- case (Dhall.TypeCheck.typeOf annot) of
-                Left err -> throwMissingImport (Imported _stack err)
-                Right _  -> return ()
-
-            return (Core.normalize loadedExpr)
-
-    let handler₀ (e :: SomeException) = do
-            {- Try to typecheck using the preferred @mapKey@/@mapValue@ fields
-               and fall back to @header@/@value@ if that fails. However, if
-               @header@/@value@ still fails then re-throw the original exception
-               for @mapKey@ / @mapValue@. -}
-            let handler₁ (_ :: SomeException) =
-                    throwMissingImport (Imported _stack e)
-
-            handle handler₁ (go "header" "value")
-
-    headersExpression' <-
-        handle handler₀ (go "mapKey" "mapValue")
-
-    return url { headers = Just (fmap absurd headersExpression') }
-
-normalizeHeaders url = return url
+    loaded <- loadWith (ImportAlt partialExpr emptyOriginHeaders)
+    headers <- liftIO (toOriginHeaders loaded)
+
+    -- return cached headers next time
+    _ <- State.modify (\state -> state { _loadOriginHeaders = return headers })
+
+    return headers
```

Comment:
This is where you will need to reset the stack to contain just the headers file.  However, it will require restructuring things a little bit because the type of `originHeadersLoader` doesn't permit it to know what path the `headersExpr` came from.  You might need to change it so that the `headersExpr` returns the original `Import` that the `Expr` came from so that `originHeadersLoader` can set the `stack` to that `Import`

Reply:
> You might need to change it so that the headersExpr returns the original Import that the Expr came from so that originHeadersLoader can set the stack to that Import

Hmm, I think I removed my ability to do this 🤦 

`headersExpr` in the normal case is `env:DHALL_HEADERS ? ~/.config/dhall/headers.dhall`. So you can't know which branch was successful until _after_ `loadWith` is called, but you need to give `loadWith` a stack.

I've pushed an alternative solution: when loading headers, we just start with the root import. I believe this is generally either `.` (for an expression) or the path to the main file (if using --file). So this should always be local.

The trick is I had to ensure reentrant calls to load the headers didn't also reset the stack, since that's how we'd detect the `Cycle` error. Unfortunately that didn't work in practice (see [this comment](https://github.com/dhall-lang/dhall-haskell/blob/dd85fbad9edb130d32b8f3ce08279c8562d67fe3/dhall/src/Dhall/Import.hs#L1071-L1077)), but conceptually that's how it's working.

I think I could remove that hack by _also_ popping the most recent item off the stack in reentrant calls (so that it doesn't look like the headers are being loaded _from_ the remote import), but it didn't seem worth that extra complexity to achieve the same result.

Reply:
Actually, I think that indicates that there is a bug in the standardized behavior.  The bug is that, as you noted, resetting the stack as the standard specifies interferes with cyclic import detection.

What if we were to change the standard to not reset the stack (like you just implemented in your most recent changes) and instead change the logic to:

```
(Δ, parent, headersPath) × Γ₀ ⊢ userHeadersExpr ⇒ userHeaders ⊢ Γ₁
```

In other words, we append the headers path to the stack (not including the current URL import that the headers will be used for) when resolving the headers file.